### PR TITLE
Add proper synchronization when access metrics.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We also need to add synchronization when access fields to ensure we see the latest updates.

Modifications:

Add synchronization when read fields that are written concurrently.

Result:

Ensure correct visibility of updated.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@buchgr @Scottmitch @trustin PTAL

**Comment by buchgr** [comment]
I don't think that's how it works. You'll need to declare these fields as volatile, or wrap the writes in a monitor. Unless, I am missing something?

**Comment by normanmaurer** [comment]
@buchgr I'm missing something ? The writes should be already in a monitor. The purpose of the PR is to just fix the "reading" as the read was not done in a monitor .

**Comment by buchgr** [comment]
@normanmaurer ahh I didn't realize the writes where in a monitor :-)

**Comment by normanmaurer** [comment]
@buchgr so all good ?

**Comment by buchgr** [comment]
yup. LGTM

**Comment by Scottmitch** [comment]
lgtm

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (daa4efcfef0f747b7ce8e77335881fbd5692bffa) and 4.1 (daa4efcfef0f747b7ce8e77335881fbd5692bffa)

File: buffer/src/main/java/io/netty/buffer/PoolArena.java [hunk 1]
```diff
@@ -446,27 +450,31 @@ public long numSmallAllocations() {
...
+    public synchronized long numNormalAllocations() {
         return allocationsNormal;
     }
 
     @Override
     public long numDeallocations() {
-        return deallocationsTiny + deallocationsSmall + allocationsNormal + deallocationsHuge.value();
+        final long deallocs;
+        synchronized (this) {
+            deallocs = deallocationsTiny + deallocationsSmall + allocationsNormal;
```
Comment (ChristianSchwarz) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: @normanmaurer 
Is this a bug?

deallocs = deallocationsTiny + deallocationsSmall + **de**allocationsNormal;

Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: @ChristianSchwarz oh shit yes! Thanks for spotting it. Fixing now.

File: buffer/src/main/java/io/netty/buffer/PoolArena.java [hunk 2]
```diff
@@ -481,8 +489,12 @@ public long numHugeDeallocations() {
     }
 
     @Override
-    public long numActiveAllocations() {
-        long val = numAllocations() - numDeallocations();
+    public  long numActiveAllocations() {
+        long val;
+        synchronized (this) {
+            val = allocationsNormal - deallocationsTiny - deallocationsSmall - allocationsNormal;
```
Comment (ChristianSchwarz) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: @normanmaurer I guess here i a bug too?
`val = allocationsNormal - deallocationsTiny - deallocationsSmall -` **de**`allocationsNormal;`

Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: @ChristianSchwarz also here is a bug... fixing as well.

Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: @ChristianSchwarz actually I think it should be:

```java
    @Override
    public  long numActiveAllocations() {
        long val = allocationsTiny.value() + allocationsSmall.value() + allocationsHuge.value() - deallocationsHuge.value();
        synchronized (this) {
            val += allocationsNormal;
            val -= deallocationsTiny + deallocationsSmall + deallocationsNormal;
        }
        return val >= 0 ? val : 0;
    }
```

WDYT ?

File: buffer/src/main/java/io/netty/buffer/PoolArena.java [hunk 3]
```diff
@@ -500,7 +512,10 @@ public long numActiveSmallAllocations() {
 
     @Override
     public long numActiveNormalAllocations() {
-        long val = numNormalAllocations() - numNormalDeallocations();
+        final long val;
+        synchronized (this) {
+            val = allocationsNormal - deallocationsNormal;
+        }
         return val >= 0 ? val : 0;
```
Comment (ChristianSchwarz) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: I think it is easier to read and understand if one would write `Math.max(val,0)` (with static import of course)..

File: buffer/src/main/java/io/netty/buffer/PoolArena.java (hunk 2 above)
Comment (ChristianSchwarz) [inline buffer/src/main/java/io/netty/buffer/PoolArena.java]: +1 looks more clean than before

**Pull request closed without merging**